magda-sleuther-duplicate-records/src/onRecordFound.ts

```diff
@@ -0,0 +1,147 @@
+import { Record } from "@magda/typescript-common/dist/generated/registry/api";
+//import unionToThrowable from "@magda/typescript-common/dist/util/unionToThrowable";
+import { DuplicateRecordsAspect, DistURL } from "./duplicateRecordsAspectDef";
+import FTPHandler from "./FtpHandler";
+
+export var mochaObject: any = { isRunning: false };
+
+export var testGroups: DuplicateRecordsAspect[];
+
+export default async function onRecordFound(
+    record: Record,
+    registry: Registry,
+    retries: number = 5,
+    baseRetryDelaySeconds: number = 1,
+    base429RetryDelaySeconds = 60,
+    ftpHandler: FTPHandler = new FTPHandler()
+) {
+    const distributions: Record[] =
+        record.aspects["dataset-distributions"] &&
+        record.aspects["dataset-distributions"].distributions;
+
+    if (!distributions || distributions.length === 0) {
+        return Promise.resolve();
+    }
+
+    // DistContainer contains only the properties we need, putting all distributions into this will help manipulate the data.
+    const distributionContainers: DistContainer[] = _.flatten(
+        distributions.map(function(distribution) {
+            if (
+                distribution.aspects &&
+                distribution.aspects["dcat-distribution-strings"]
+            ) {
+                let distContainer: DistContainer[] = [];
+                distContainer.push({
+                    url: {
+                        url: null,
+                        type: null
+                    },
+                    id: null
+                });
+                distContainer.push({
+                    url: {
+                        url: null,
+                        type: null
+                    },
+                    id: null
+                });
+
+                
+                distContainer[0].url = {
+                    url: distribution.aspects["dcat-distribution-strings"]
+                        .accessURL as string,
+                    type: "accessURL" as "accessURL"
+                } || { type: "none" };
+
+                distContainer[1].url = {
+                    url: distribution.aspects["dcat-distribution-strings"]
+                        .downloadURL as string,
+                    type: "downloadURL" as "downloadURL"
+                } || { type: "none" };
+
+                distContainer[0].id = distContainer[1].id = distribution.id;
+
+                // adding 2 instances of distcontainer with the same url and same id just adds garbage to the db, so delete it if its the case.
+                if (
+                    (distribution.aspects["dcat-distribution-strings"]
+                        .accessURL as string) ===
+                    (distribution.aspects["dcat-distribution-strings"]
+                        .downloadURL as string)
+                ) {
+                    distContainer.pop();
+                }
+
+                return distContainer;
+            }
+
+            throw new Error(
```

Comment:
Even though it's not strictly necessary, it'd be good to put this in an `else` block so it's obvious that it's not supposed to be on the main path of execution.